src/clj/rems/api/applications_v2.clj

```diff
@@ -0,0 +1,869 @@
+      (is (= #{:foo :bar :gazonk}
+             (user-permissions application "user")))
+      (is (= nil
+             (user-permissions application "wrong user"))))))
+
+;;;; v2 API, pure application state based on application events
+
+(defmulti ^:private application-view-specific
+  "See `application-view`"
+  (fn [_application event] (:event/type event)))
+
+(defmethod application-view-specific :application.event/created
+  [application event]
+  (-> application
+      (assoc :application/id (:application/id event)
+             :application/created (:event/time event)
+             :application/modified (:event/time event)
+             :application/applicant (:event/actor event)
+             :application/resources (map (fn [resource]
+                                           {:catalogue-item/id (:catalogue-item/id resource)
+                                            :resource/ext-id (:resource/ext-id resource)})
+                                         (:application/resources event))
+             :application/licenses (map (fn [license]
+                                          {:license/id (:license/id license)
+                                           :license/accepted false})
+                                        (:application/licenses event))
+             :application/events []
+             :form/id (:form/id event)
+             :form/fields []
+             :workflow/id (:workflow/id event)
+             :workflow/type (:workflow/type event)
+             ;; TODO: or would :workflow.dynamic/state be more appropriate?
+             :workflow/state :rems.workflow.dynamic/draft ; TODO: other workflows
+             :workflow.dynamic/handlers (:workflow.dynamic/handlers event))
+      (set-role-permissions {:applicant #{::dynamic/add-member
+                                          ::dynamic/save-draft
+                                          ::dynamic/submit}})))
+
+(defn- set-accepted-licences [licenses accepted-licenses]
+  (map (fn [license]
+         (assoc license :license/accepted (contains? accepted-licenses (:license/id license))))
+       licenses))
+
+(defmethod application-view-specific :application.event/draft-saved
+  [application event]
+  (-> application
+      (assoc :application/modified (:event/time event)
+             :form/fields (map (fn [[field-id value]]
+                                 {:field/id field-id
+                                  :field/value value})
+                               (:application/field-values event)))
+      (update :application/licenses set-accepted-licences (:application/accepted-licenses event))))
+
+(defmethod application-view-specific :application.event/member-added
+  [application event]
+  application)
+
+(defmethod application-view-specific :application.event/submitted
+  [application event]
+  (-> application
+      (assoc :workflow/state ::dynamic/submitted)
+      (set-role-permissions {:applicant #{::dynamic/add-member}
```

Comment:
These are not up to date anymore.

`add-member` is only for the handler in specific states and `invite-member` is both for the applicant and handler in specific states.

Reply:
Any incompatibility will be fixed by the second step in #852 by running the tests for the old implementation against this new implementation. That's outside the scope of this PR, because it will require changes to the existing code; in particular the dynamic workflow tests need to be refactored to be purely event-based.

Reply:
Then there should at least be a TODO or maybe just assert that always throws that the problem is obvious later?

Reply:
::dynamic/add-member has now been removed

Reply:
The TODO note is in 852